storage: only acquire read lock when checking for in-progress merge

benesch opened the pull request:
Replica.mu is a reader-writer lock, so we can be slightly more efficient
when we check to see whether a merge is in progress by only acquiring a
read lock.

Release note: None

benesch committed:
storage: only acquire read lock when checking for in-progress merge

Replica.mu is a reader-writer lock, so we can be slightly more efficient
when we check to see whether a merge is in progress by only acquiring a
read lock.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/27400)
<!-- Reviewable:end -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27400#-:-LH8wSqY7vE5ixsob9Xa:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

benesch committed:
storage: avoid race in TestStoreRangeMergeWithData

Request filters are called from multiple threads, so sync/atomic to
update the retries counter in TestStoreRangeMergeWithData to avoid
races.

Fix #27401.
Fix #27403.
Fix #27405.

Release note: None

benesch commented:
Updated to also fix the race that's failing https://github.com/cockroachdb/cockroach/issues/27403 and friends. PTAQL!

nvanbenschoten commented on pkg/storage/client_merge_test.go:
```diff
@@ -186,25 +187,25 @@ func TestStoreRangeMergeWithData(t *testing.T) {
 			for _, req := range ba.Requests {
 				if et := req.GetEndTransaction(); et != nil && et.InternalCommitTrigger.GetMergeTrigger() != nil {
-					retries--
+					atomic.AddInt64(&retries, -1)
```
Do we want some kind of CAS approach here to avoid letting `retries` drop below 0? Does that matter?

bdarnell reviewed (approved):
> Reviewed 1 of 1 files at r1, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27400)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

benesch commented:
bors r=bdarnell,tschottdorf,nvanbenschoten
